c_common/models/extra_monitor_support/extra_monitor_support.c

```diff
@@ -882,113 +892,168 @@ void the_dma_complete_read_missing_seqeuence_nums() {
             retransmission_dma_read();
         }
     } else {
-        // get next seq num to regenerate
+        // get next sequence number to regenerate
         missing_seq_num_being_processed = (uint32_t)
             retransmit_seq_nums[position_in_read_data];
+        //io_printf(IO_BUF, "dealing with seq num %d \n",
+         //         missing_seq_num_being_processed);
         if (missing_seq_num_being_processed != END_FLAG) {
+
             // regenerate data
             position_in_store = missing_seq_num_being_processed * (
-        	    ITEMS_PER_DATA_PACKET - SEQUENCE_NUMBER_SIZE);
-            read(DMA_TAG_RETRANSMISSION_READING, 1,
-        	    (ITEMS_PER_DATA_PACKET - SEQUENCE_NUMBER_SIZE));
+                ITEMS_PER_DATA_PACKET - SEQUENCE_NUMBER_SIZE);
+            uint32_t left_over_portion =
+                ((bytes_to_read_write / WORD_TO_BYTE_MULTIPLIER) -
+                position_in_store);
+
+                //io_printf(IO_BUF, "for seq %d, pos = %d, left %d\n",
+                //missing_seq_num_being_processed, position_in_store,
+                //left_over_portion);
+
+            if (left_over_portion <
+                    (ITEMS_PER_DATA_PACKET - SEQUENCE_NUMBER_SIZE)){
+                retransmitted_seq_num_items_read = left_over_portion + 1;
+                read(DMA_TAG_RETRANSMISSION_READING, 1, left_over_portion);
+            }else
+            {
+                retransmitted_seq_num_items_read = ITEMS_PER_DATA_PACKET;
+                read(DMA_TAG_RETRANSMISSION_READING, 1,
+                    (ITEMS_PER_DATA_PACKET - SEQUENCE_NUMBER_SIZE));
+            }
         } else {        // finished data send, tell host its done
             data_speed_up_send_end_flag();
+            in_re_transmission_mode = false;
+            missing_sdp_seq_num_sdram_address = NULL;
+            position_in_read_data = 0;
+            position_for_retransmission = 0;
+            number_of_missing_seq_nums_in_sdram = 0;
         }
     }
 }
 
-//! \brief dma complete callback for have read missing seq num data
+//! \brief DMA complete callback for have read missing sequence number data
 void dma_complete_reading_retransmission_data() {
     //log_info("just read data for a given missing sequence number");
 
-    // set seq number as first element
+    // set sequence number as first element
     data_to_transmit[transmit_dma_pointer][0] =
-	    missing_seq_num_being_processed;
+        missing_seq_num_being_processed;
+
+	if (missing_seq_num_being_processed > max_seq_num) {
+	    io_printf(
+	        IO_BUF, "Got some shitty seq num here. max is %d and got %d \n",
```

Comment:
Unprofessional language in a message that might (but hopefully won't! 😉) get seen by users.